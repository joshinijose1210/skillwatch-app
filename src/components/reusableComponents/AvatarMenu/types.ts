export interface AvatarMenuProps {
    open: boolean;
    setOpen: React.Dispatch<React.SetStateAction<boolean>>;
}
