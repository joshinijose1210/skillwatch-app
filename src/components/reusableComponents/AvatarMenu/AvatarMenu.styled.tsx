import { Box, Text } from '@medly-components/core';
import { defaultTheme } from '@medly-components/theme';
import styled, { css } from 'styled-components';

export const AvatarMenu = styled.div`
    z-index: 10;
`;

export const Wrapper = styled.ul`
    list-style: none;
    margin-left: -4rem;
`;

export const DropdownMenu = styled.div`
    position: absolute;
    top: 65px;
    right: 20px;
    background-color: ${defaultTheme.colors.white};
    border-radius: 8px;
    padding: 10px;
    width: 300px;
    border: 1px solid ${defaultTheme.colors.grey[300]};
    box-shadow: 5px 5px 20px ${defaultTheme.colors.grey[400]};

    &::before {
        content: '';
        position: absolute;
        top: -5px;
        right: 20px;
        background: ${defaultTheme.colors.white};
        transform: rotate(45deg);
    }

    &.active {
        opacity: 1;
        visibility: visible;
        transform: translateY(0);
        transition: 500ms ease;
    }

    &.inactive {
        opacity: 0;
        visibility: hidden;
        transform: translateY(-20px);
        transition: 500ms ease;
    }
`;

export const MenuTrigger = styled.div`
    overflow: hidden;
    cursor: pointer;
    background-color: ${defaultTheme.colors.white};
`;

export const AvatarList = styled.li`
    height: 80px;
    width: 80px;
    margin: auto;
    margin-bottom: 1rem;
`;

export const EmailList = styled.li<{ isPopoverClosed: boolean }>`
    text-align: center;

    #medly-popover-wrapper h5 {
        overflow: hidden;
        text-overflow: ellipsis;
        max-width: 300px;
    }

    ${({ isPopoverClosed }) =>
        isPopoverClosed &&
        css`
            #medly-popover-wrapper #medly-popover-popup {
                display: none;
            }
        `}
`;

export const SignOutButtonList = styled.li`
    padding-top: 2rem;
    display: flex;
    justify-content: center;
    align-items: center;
    border-top: 1px solid ${({ theme }) => theme.customColors.borderColor};
    margin-top: 1rem;
`;

export const StyledLabel = styled(Text).attrs({
    textVariant: 'h5',
    textAlign: 'center',
    textWeight: 'Medium'
})`
    span {
        font-weight: normal;
    }
`;

export const StyledBox = styled(Box)`
    margin: 0;
    padding: 0;
    row-gap: 5px;
    flex-direction: column;
`;
