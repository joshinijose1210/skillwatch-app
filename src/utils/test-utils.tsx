import { ToastContainer } from '@medly-components/core';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { store as rtkStore, setupStore } from '@slice';
import { RenderOptions, RenderResult, render } from '@testing-library/react';
import { defaultTheme } from '@theme';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import reduxMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { userMockState } from '../mocks/userMockState';
import { initialState } from '../slice';

interface CustomOptions extends RenderOptions {
    initialState?: any;
    initialEntries?: any;
}

interface WithThemeProviderProps {
    children?: React.ReactNode;
}

const mockStore = reduxMockStore(),
    store = mockStore({
        ...initialState,
        user: { ...initialState.user, groups: ['admin'] }
    });

const WithThemeProvider: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <ThemeProvider theme={defaultTheme}>
        <>
            <ToastContainer position="top-end" />
            {children}
        </>
    </ThemeProvider>
);

const WithStore: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <WithThemeProvider>{children}</WithThemeProvider>
    </Provider>
);

const WithRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <MemoryRouter>
        <WithThemeProvider>{children}</WithThemeProvider>
    </MemoryRouter>
);

const WithStoreAndRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <MemoryRouter>
            <WithThemeProvider>{children}</WithThemeProvider>
        </MemoryRouter>
    </Provider>
);

export const RTKRender = (ui: React.ReactElement, { initialState = userMockState, ...renderOptions }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);

    const RtkWrapper: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
        <Provider store={store}>
            <MemoryRouter initialEntries={renderOptions.initialEntries}>
                <WithThemeProvider>{children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
    return { rtkStore, ...render(ui, { wrapper: RtkWrapper, ...renderOptions }) };
};
const customRender = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithThemeProvider, ...options });

export const renderWithStore = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStore, ...options });

export const renderWithRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithRouter, ...options });

export const renderWithStoreAndRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStoreAndRouter, ...options });

export const wrapper = (props: { children: React.ReactNode }, { initialState = userMockState }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);
    return (
        <Provider store={store}>
            <MemoryRouter>
                <WithThemeProvider>{props.children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render, store };
