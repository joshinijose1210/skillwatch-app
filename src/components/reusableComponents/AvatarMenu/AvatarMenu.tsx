import apiUrls from '@constants/apiUrls';
import { addToast, Avatar, Text, Popover, Button } from '@medly-components/core';
import { useAppDispatch, useAppSelector } from '@slice';
import { updateReviewCycleFilter } from '@slice/reviewCycleFilter';
import { removeUser } from '@slice/user';
import { fetch } from '@utils';
import { FC, useEffect, useRef, useMemo } from 'react';
import * as Styled from './AvatarMenu.styled';
import { AvatarMenuProps } from './types';

export const AvatarMenu: FC<AvatarMenuProps> = ({ open, setOpen }: AvatarMenuProps) => {
    const { departmentName, designationName, firstName, lastName, email, roleName, phoneNumber, teamName, id, employeeId } = useAppSelector(
        state => state.user
    );

    const dispatch = useAppDispatch();

    const menuRef = useRef();
    const emailTextRef = useRef<HTMLElement>(null);

    useEffect(() => {
        const handler = (e: any) => {
            if ((menuRef.current as any) && !(menuRef.current as any).contains(e.target)) {
                setOpen(false);
            }
        };

        document.addEventListener('mousedown', handler);

        return () => {
            document.removeEventListener('mousedown', handler);
        };
    }, [menuRef, setOpen]);

    const handleLogout = async () => {
        const refreshToken = JSON.parse(localStorage.getItem('token') as string)?.refresh_token;

        const { response, error } = await fetch({
            url: apiUrls.logout,
            method: 'POST',
            data: {
                id,
                refreshToken
            }
        });
        if (response?.data) {
            dispatch(removeUser());
            dispatch(updateReviewCycleFilter({ reviewCycleList: [] }));
        }
        if (error) {
            addToast({ variant: 'error', message: 'Logout failed, please try again' });
        }
    };

    const isPopoverClosed = useMemo(
        () => (emailTextRef?.current && emailTextRef?.current.getBoundingClientRect().width >= 300 ? false : true),
        [open, emailTextRef.current]
    );

    return (
        <Styled.AvatarMenu ref={menuRef as any}>
            <Styled.MenuTrigger onClick={() => setOpen(!open)}>
                <Avatar size="M">
                    {firstName[0]}
                    {lastName[0]}
                </Avatar>
            </Styled.MenuTrigger>

            <Styled.DropdownMenu className={`${open ? 'active' : 'inactive'}`}>
                <Styled.Wrapper>
                    <Styled.AvatarList>
                        <Avatar
                            style={{
                                height: '100%',
                                width: '100%'
                            }}
                        >
                            <Text style={{ margin: '2rem 2rem', fontSize: '3rem', fontWeight: '1rem' }}>
                                {firstName[0]?.toUpperCase()}
                                {lastName[0]?.toUpperCase()}
                            </Text>
                        </Avatar>
                    </Styled.AvatarList>
                    <Styled.StyledBox>
                        <Text textVariant="h4" textAlign="center">
                            {firstName} {lastName} {employeeId && `(${employeeId})`}
                        </Text>

                        {departmentName && (
                            <Styled.StyledLabel>
                                Department: <span>{departmentName}</span>
                            </Styled.StyledLabel>
                        )}

                        {teamName && (
                            <Styled.StyledLabel>
                                Team: <span>{teamName}</span>
                            </Styled.StyledLabel>
                        )}

                        {designationName && (
                            <Styled.StyledLabel>
                                Designation: <span>{designationName}</span>
                            </Styled.StyledLabel>
                        )}

                        {roleName && (
                            <Styled.StyledLabel>
                                Role: <span>{roleName}</span>
                            </Styled.StyledLabel>
                        )}

                        <Styled.EmailList isPopoverClosed={isPopoverClosed}>
                            <Popover>
                                <Styled.StyledLabel ref={emailTextRef}>{email}</Styled.StyledLabel>
                                <Popover.Popup withArrow placement="left">
                                    {email}
                                </Popover.Popup>
                            </Popover>
                        </Styled.EmailList>

                        {phoneNumber && <Styled.StyledLabel>{phoneNumber}</Styled.StyledLabel>}
                    </Styled.StyledBox>
                    <Styled.SignOutButtonList>
                        <Button variant="outlined" onClick={handleLogout}>
                            Sign Out
                        </Button>
                    </Styled.SignOutButtonList>
                </Styled.Wrapper>
            </Styled.DropdownMenu>
        </Styled.AvatarMenu>
    );
};
