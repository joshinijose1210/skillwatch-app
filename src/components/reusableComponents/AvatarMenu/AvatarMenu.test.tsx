import { RTKRender } from '@utils/test-utils';
import { AvatarMenu } from './AvatarMenu';

describe('AvatarMenu component', () => {
    it('should take snapshot', () => {
        const { container } = RTKRender(<AvatarMenu open={false} setOpen={jest.fn()} />);
        expect(container).toMatchSnapshot();
    });
});
