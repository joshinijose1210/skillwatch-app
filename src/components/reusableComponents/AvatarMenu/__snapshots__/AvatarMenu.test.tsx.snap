// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarMenu component should take snapshot 1`] = `
<div>
  <div
    class="sc-1298oxm-0 dQDoHN"
  >
    <div
      class="sc-1298oxm-3 kwZxBE"
    >
      <div
        class="sc-1dzehji-0 hjLOhC"
        defaults="[object Object]"
      >
        <strong
          class="sc-1ntgl3l-0 fzLUlA"
          textalign="initial"
          textcolor="inherit"
          textweight="Strong"
        >
          A
        </strong>
        <strong
          class="sc-1ntgl3l-0 fzLUlA"
          textalign="initial"
          textcolor="inherit"
          textweight="Strong"
        >
          D
        </strong>
      </div>
    </div>
    <div
      class="sc-1298oxm-2 ekJHDu inactive"
    >
      <ul
        class="sc-1298oxm-1 jVMYcc"
      >
        <li
          class="sc-1298oxm-4 giVItn"
        >
          <div
            class="sc-1dzehji-0 hjLOhC"
            defaults="[object Object]"
            style="height: 100%; width: 100%;"
          >
            <span
              class="sc-1ntgl3l-0 hGagSU"
              style="margin: 2rem 2rem; font-size: 3rem; font-weight: 1rem;"
              textalign="initial"
              textcolor="inherit"
            >
              A
              D
            </span>
          </div>
        </li>
        <div
          bg="white"
          borderradius="2"
          class="sc-12ecrgm-0 dAEUnm sc-1298oxm-8 Hlhle"
          color="black"
          display="flex"
          font-variant="body2"
          px="4"
          py="4"
        >
          <h4
            class="sc-1ntgl3l-0 igfIhV"
            textalign="center"
            textcolor="inherit"
            textvariant="h4"
          >
            ABC
             
            DEF
             
            (SR0022)
          </h4>
          <h5
            class="sc-1ntgl3l-0 dBSnDZ sc-1298oxm-7 kNgBfF"
            textalign="center"
            textcolor="inherit"
            textvariant="h5"
            textweight="Medium"
          >
            Team: 
            <span>
              BA
            </span>
          </h5>
          <h5
            class="sc-1ntgl3l-0 dBSnDZ sc-1298oxm-7 kNgBfF"
            textalign="center"
            textcolor="inherit"
            textvariant="h5"
            textweight="Medium"
          >
            Designation: 
            <span>
              BA - Traineee
            </span>
          </h5>
          <h5
            class="sc-1ntgl3l-0 dBSnDZ sc-1298oxm-7 kNgBfF"
            textalign="center"
            textcolor="inherit"
            textvariant="h5"
            textweight="Medium"
          >
            Role: 
            <span>
              Manager
            </span>
          </h5>
          <li
            class="sc-1298oxm-5 jElgbs"
          >
            <div
              class="sc-1ep6kh8-0 hrIcnd"
              id="medly-popover-wrapper"
            >
              <h5
                class="sc-1ntgl3l-0 dBSnDZ sc-1298oxm-7 kNgBfF"
                textalign="center"
                textcolor="inherit"
                textvariant="h5"
                textweight="Medium"
              >
                abc@scalereal.com
              </h5>
            </div>
          </li>
          <h5
            class="sc-1ntgl3l-0 dBSnDZ sc-1298oxm-7 kNgBfF"
            textalign="center"
            textcolor="inherit"
            textvariant="h5"
            textweight="Medium"
          >
            +917060753645
          </h5>
        </div>
        <li
          class="sc-1298oxm-6 bKQgdL"
        >
          <button
            class="sc-1k2druj-0 dsEKPA"
            edges="square"
            type="button"
            variant="outlined"
          >
            <span
              class="sc-1ntgl3l-0 eOsHHD"
              textalign="initial"
              textcolor="inherit"
              textvariant="button1"
            >
              Sign Out
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
